### PR TITLE
Supervised Contrastive Loss

### PR DESCRIPTION
**Pull request by waleeattia**
#### Reference issue
[#426](https://github.com/neurodata/ProgLearn/issues/426)

#### Type of change
Implementing supervised contrastive loss
Adding plotting script to compare accuracies and transfer efficiencies

#### What does this implement/fix?
Implementing contrastive loss explicitly learns the progressive learning network transformer by penalizing samples of different classes that are close to one another. The new script enables two dnn algorithms to be compared by plotting the difference between their accuracies and transfer efficiencies. The accuracy of the supervised contrastive loss version improves by 6 percent compared to the PL network with categorical cross entropy. 

#### Additional information
NDD 2021

### CODE REVIEW

**Comment by jdey4** [comment]
@rflperry Does this PR help your query about contrastive loss?

**Comment by rflperry** [comment]
Yeah seems like it matches [my results here](https://github.com/rflperry/ProgLearn/blob/contrast-loss/benchmarks/cifar_exp/result/figs/benchmark_contrast4.pdf) that the transfer ability goes down, which I find interesting but the reason why I'm still a bit intrigued by. Not really worth adding if just always worse? I forget why I had multiple different results with different labels.

**Comment by rflperry** [comment]
My takeaways/summary:
- Since the decider is k-Nearest Neighbors, we want the learned (penultimate) representation to place samples of the same class close together.
- Contrastive loss learns representations that are close together, and this is validated as we see higher accuracy from our kNN classifier. Softmax worked, but wasn't explicitly tuned to learn what we wanted (in a way, the best loss would be a function of the network and decider together).
- One slightly odd thing is that the difference in accuracy is non-monotonic (i.e. goes down then up). Maybe just a result of not running enough simulations?
- Despite the accuracy going up, the transfer efficiencies are slightly worse. I'm a bit fuzzy on the details of the transfer efficiency metric, but potentially the learned embeddings are not good for OOD performance (this has been observed in various learned embedding algorithms like tSNE I believe)

**Review by PSSF23 — CHANGES_REQUESTED** [review]
@waleeattia Put figure in correct folder & save in `pdf` format.

**Comment by waleeattia** [comment]
@PSSF23 fixed!

**Review by PSSF23 — CHANGES_REQUESTED** [review]
Remove commented code & unnecessary prints. After these LGTM.

**Comment by waleeattia** [comment]
@PSSF23 Perfect, just made those changes. Thank you!

**Review by PSSF23 — COMMENTED** [review]
Still some commented code remaining in `benchmarks/cifar_exp/plot_compare_two_algos.py` @waleeattia

**Comment by waleeattia** [comment]
@PSSF23 Sorry I missed that, it should be good now.

**Review by PSSF23 — APPROVED** [review]
